File: 250428/2명의 도둑/two-thieves.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

int n, m, c;
int weight[11][11];
int sum = INT_MIN;

void func(int row, int col, int num, int curr, int val) {
    if(col > n || curr > c) return;

    if (num > m) {
        sum = max(sum, val);
        return;
    }

    if (curr + weight[row][col] <= c) {
        func(row, col + 1, num + 1, curr + weight[row][col], val + weight[row][col] * weight[row][col]);
    }
    func(row, col + 1, num + 1, curr, val);
}

int main() {
    cin >> n >> m >> c;

    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= n; j++) {
            cin >> weight[i][j];
        }
    }
    int result = 0;
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= n; j++) {
            func(i, j, 1, 0, 0);
            int first_thief = sum;
            sum = INT_MIN;
            // if(first_thief == 64) cout << i << " and " << j << '\n';
            for (int k = 1; k <= n; k++) {
                for (int l = 1; l <= n; l++) {
                    if (i == k && !(l + m - 1 < j || l > j + m - 1)) continue;
                    sum = INT_MIN;
                    func(k, l, 1, 0, 0);
                    int second_thief = sum;
                    // cout << "sum is " << sum << '\n';
                    result = max(result, first_thief + second_thief);
                }
            }
        }
    }
    cout << result;
    return 0;
}

File: 250807/2명의 도둑/two-thieves.py
n, m, c = map(int, input().split())
weight = [list(map(int, input().split())) for _ in range(n)]
visited = [False] * n
selected = []
ans = 0
# 행을 선택한다. visited(col) 표시
# 행에서의 가능한 n - 1경우의 수를 수행한다.
# 각각의 수행마다 또 다른 도둑의 경우의 수를 백트래킹 수행

def print_output():
    print(*selected)

def get_weight(row, col, length):
    return sum(x**2 for x in weight[row][col:col + length])

def backtracking(curr):
    if curr > c:
        # print_output()
        global ans
        ans = max(ans, curr)
        return
    else:
        for i in range(n):
            for j in range(n - m):
                if not visited[j]:
                    for in range(j,j + m):
                        total weight = 0

                    selected.extend(weight[i][j:j+m])
                    visited[j:j+m] = [True] * m
                    backtracking(curr + get_weight(i, j, m))
                    visited[j:j+m] = [False] * m
                    selected.extend(weight[i][j:j+m])
                    del selected[-m:]

backtracking(0)
print(ans)